[spring-ecommerce-order - Step-1] Submit AnnLvu's mission

AnnLvu opened the pull request:
## 🔄 Spring E-commerce-order - Step 1 PR Overview
Hi PK 👋,

This PR represents **Step 1** of our Spring E-commerce platform improvements. The main focus of this step was to **migrate from JDBC to Spring Data JPA**, implement **pagination**, and **add support for product options**.

### ✅ Features Implemented

#### 🔁  JDBC → Spring Data JPA
- Replaced all repository logic with Spring Data JPA interfaces
- Removed direct SQL/JdbcTemplate usage
- Introduced proper entity mapping using annotations:
  - `@Entity`, `@Id`, `@OneToMany`, `@ManyToOne`

#### 🧩 Entity Relationships
- Defined entities: `Product`, `Option`, `Cart`, `CartProduct`, `User` (Member/Admin)
- Maintained bidirectional relationships where necessary
- Added conversion logic between DTOs and entities (`toDTO()`, `toEntity()`)

#### 📄 Pagination Support
- Implemented pagination for product listings (both for guest and admin views)
- Controlled via `page` and `perPage` query parameters
- Default values configured and reusable

#### ⚙️ Product Options
- Each product now supports a list of options 
- Options include `name`, `price`, `quantity`, `imageUrl`
- Full CRUD support for options 
- Validations:
  - Unique option name per product
  - Quantity ≥ 0, valid price, etc.
- `PATCH` support to partially update option fields

#### 🔄 Question
In several places (for example: `Product → Option`, `Cart → CartProduct`), I'm using:

```kotlin
@OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true)

```
This allows managing child entities directly from the parent, including creation and removal.

**Questions:**

- Is it appropriate to use `CascadeType.ALL` in these contexts, or would it be better to specify only the required cascade types (PERSIST, MERGE, etc.) more explicitly?

- I'm also interested in whether I've correctly understood and applied the JPA relationships themselves (e.g., mappedBy, @JoinColumn, orphanRemoval, etc.). Are there any aspects of the mapping you think could be improved?


### ❗️Important:
To test the project, you need to create a .env file at the same level as build.gradle.kts.
The file should contain the following parameters:

```kotlin
JWT_SECRET=
JWT_TIME=

```
And put your own secret values there

Thanks a lot for your time!

MehrShahbaz committed:
init: copy and paste previous mission

MehrShahbaz committed:
feat: add JPA to project

MehrShahbaz committed:
refactor: remove .gitkeep file

MehrShahbaz committed:
refactor(Product, ProductRepository): convert Product to JPA ent

MehrShahbaz committed:
refactor(Cart, CartRepository): convert Cart to JPA entity

MehrShahbaz committed:
refactor(CartProduct, CartProductRepository): convert CartProduct to JPA entity

MehrShahbaz committed:
refactor(CartStatistics, CartStatisticsRepository): convert CartStatistics to JPA entity

MehrShahbaz committed:
refactor(User, UserRepository): convert User to JPA entity

MehrShahbaz committed:
docs(README.md): add entity

MehrShahbaz committed:
refactor(LoginService, UserRepository): integrate JPA
- add new method `findByEmailAndPassword`

MehrShahbaz committed:
fix(Repository): fix import issues

MehrShahbaz committed:
refactor(MemberAuthService): convert to JPA repository

MehrShahbaz committed:
refactor(ProductDTO): remove description field

MehrShahbaz committed:
refactor(GuestProductService): change to JPA repo
- ProductResponseDTO to handle the response
- util method to convert entity to responseDTO

MehrShahbaz committed:
refactor(Mapper): remove mapper dir

MehrShahbaz committed:
refactor(LoginMemberArgumentResolver, CartController): change return user entity

MehrShahbaz committed:
refactor(CartProductResponseDTO): remove unused column

MehrShahbaz committed:
refactor(GuestProductService): handle JPA entity

MehrShahbaz committed:
refactor(CartService): handle JPA entity

MehrShahbaz committed:
refactor(AdminProductService): handle JPA entity

MehrShahbaz committed:
refactor(entity): add open to entities

MehrShahbaz committed:
refactor(TopAddedProductsDTO): rename values

MehrShahbaz committed:
refactor(CartStatisticsService, CartStatisticsRepository): use JPQL instead of SQL
- add method `findTopProducts`
- add method `findActiveUsersSince`

MehrShahbaz committed:
refactor(entity): remove old entities

MehrShahbaz committed:
refactor(entity): rename jpaEntity to entity

MehrShahbaz committed:
refactor(dto): removed unused DTO's

MehrShahbaz committed:
refactor(AdminProductController): change return type for controller

MehrShahbaz committed:
refactor(BaseAuthInterceptor): handle throw message

MehrShahbaz committed:
refactor: dir change for entity

MehrShahbaz committed:
refactor(LoginRequest): add validation for login request

MehrShahbaz committed:
refactor(GuestProductController): change return type

MehrShahbaz committed:
fix(CartProduct): fix index key

MehrShahbaz committed:
refactor(application.properties): add JPA db dependency

MehrShahbaz committed:
refactor(build.gradle.kts): add all open for spring boot

MehrShahbaz committed:
refactor(entity): remove open keyword

MehrShahbaz committed:
refactor: remove unused files

MehrShahbaz committed:
style: kt-lint

MehrShahbaz committed:
refactor: remove unused column

MehrShahbaz committed:
test: fix

MehrShahbaz committed:
refactor(Product): remove unique

MehrShahbaz committed:
refactor(CartStatisticsRepository): add limit

MehrShahbaz committed:
refactor(AdminStatisticsService): add limit via PageRequest

MehrShahbaz committed:
refactor(cartStatisticsRepository): lint

MehrShahbaz committed:
refactor: rearrange dir

MehrShahbaz committed:
refactor: change id position in models

MehrShahbaz committed:
feat(GlobalExceptionHandler): handle IllegalArgumentException

MehrShahbaz committed:
refactor(ProductPatchDTO): remove unused field

MehrShahbaz committed:
refactor(Product): fix field mapping

MehrShahbaz committed:
refactor(guestController): add pagination

MehrShahbaz committed:
feat(AdminProductController): add pagination
- product repo extension function to get paginated products

MehrShahbaz committed:
refactor(ProductUtil): fix dir

MehrShahbaz committed:
refactor: change List<ProductResponseDTO> -> Page<ProductResponseDTO>

MehrShahbaz committed:
refactor(Cart): remove unused method

MehrShahbaz committed:
feat(CartController): implement clear cart

MehrShahbaz committed:
feat(WishList): implement WishListController.kt

MehrShahbaz committed:
refactor(DuplicateProductNameException): make it generic

MehrShahbaz committed:
refactor(CartService): add check for quantity

MehrShahbaz committed:
refactor(WishList): pagination

MehrShahbaz committed:
refactor(WishListProduct): add unique constraint

MehrShahbaz committed:
refactor(User, Cart): convert to uni directional

MehrShahbaz committed:
refactor(model): remove table declaration

MehrShahbaz committed:
refactor(CartStatistic): rename

MehrShahbaz committed:
refactor(BaseAuthInterceptor): fix

MehrShahbaz committed:
refactor(CartControllerTest): use global product

MehrShahbaz committed:
refactor(CartServiceTest): rename

MehrShahbaz committed:
refactor(MemberAuthControllerTest): remove unused DI

MehrShahbaz committed:
refactor(MemberAuthService): convert to uni directional

MehrShahbaz committed:
refactor(WebConfig): add wishlist to protected route

MehrShahbaz committed:
test(CartControllerTest): test

MehrShahbaz committed:
refactor(WishList): remove wishlist

MehrShahbaz committed:
feat(Options): Option Entity and DTO

MehrShahbaz committed:
feat(Options): service and controller

MehrShahbaz committed:
feat(Options): changes for option

MehrShahbaz committed:
feat(Options): changes for option for tests

MehrShahbaz committed:
refactor(User): remove wishlist

AnnLvu committed:
refactor(CartProductControllerTest): fix test
- add table name in CartProduct
- fix columnNames

AnnLvu committed:
style(AdminProductControllerTest): use ktlintFormat

AnnLvu committed:
refactor(CartServiceTest): fix CartServiceTest

AnnLvu committed:
refactor(ProductServiceTest): fix ProductServiceTest

AnnLvu committed:
style(): use ktlintformat

MehrShahbaz committed:
refactor: fix

MehrShahbaz committed:
refactor: move imageUrl to Option

MehrShahbaz committed:
refactor(DTO): fix DTO

MehrShahbaz committed:
refactor(AdminProductService): fix options

MehrShahbaz committed:
refactor(CartService): Changes for cart service

MehrShahbaz committed:
refactor(OptionRepository): Remove product dependency

MehrShahbaz committed:
refactor(toProductDTO): fix

MehrShahbaz committed:
refactor(DTO): fix

MehrShahbaz committed:
test: fix

MehrShahbaz committed:
refactor(OptionDTO.toEntity): fix

MehrShahbaz committed:
test(CartControllerTest): test

MehrShahbaz committed:
refactor(CartServiceTest): moved to end to end test

MehrShahbaz committed:
refactor: fix

MehrShahbaz committed:
test: cart

AnnLvu committed:
docs(README): update feature list

AnnLvu committed:
feat: add file for testing

pkeugine commented on src/main/resources/application.properties:
```diff
@@ -1 +1,13 @@
+spring.jpa.hibernate.ddl-auto=create
+spring.jpa.show-sql=true
+spring.jpa.properties.hibernate.format_sql=true
+#spring.jpa.open-in-view=false
```
Although these aren’t application code files, some are missing trailing newlines.
Let’s add them and make sure our IDE is set to include them automatically. 👍

This helps avoid unnecessary diffs in git.
Otherwise, teammates will keep seeing these lines as "changed," which clutters the git history.

AnnLvu replied:
Hi PK! :relaxed:
I’ve enabled the "Ensure line feed at file end on Save" option in IntelliJ IDEA.
Now, the IDE automatically adds a newline at the end of files when saving, which helps avoid unnecessary diffs in Git. Also, I  removed unnecessary property.

pkeugine commented on src/main/resources/application.properties:
```diff
@@ -1 +1,13 @@
+spring.jpa.hibernate.ddl-auto=create
+spring.jpa.show-sql=true
+spring.jpa.properties.hibernate.format_sql=true
+#spring.jpa.open-in-view=false
```
I think it would be better to remove the property instead of commenting it out if we're not going to use it.

Although it's interesting that you've tried this out, what's this `open-in-view` setting? 🙂

AnnLvu replied:
Yeah, we actually talked about this with Jason - he said it’s usually better to keep `spring.jpa.open-in-view=true` in production, since it helps avoid `LazyInitializationException` when entities are accessed outside a transaction.
We were just experimenting locally with turning it off and totally forgot to remove the commented line afterwards.
Cleaned it up now - thanks for catching it! 😄

pkeugine commented on src/main/kotlin/ecommerce/dto/products/OptionDTO.kt:
```diff
@@ -0,0 +1,29 @@
+import jakarta.validation.constraints.Size
+import org.hibernate.validator.constraints.Range
+
+class OptionDTO(
```
Class names should follow [the general camel case format, even when acronyms are included](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case).
In this case, `OptionDTO` could be changed to `OptionDto`. Why don't we change it? :)

Same review applies to function names as well!

AnnLvu replied:
Thanks for pointing that out. I’ve gone ahead and renamed all `*DTO` to `*Dto` across the project to follow the proper CamelCase convention.
Sorry about that - I initially kept DTO in all caps because I considered it an abbreviation and thought it would be more readable that way.

pkeugine commented on src/main/kotlin/ecommerce/model/Cart.kt:
```diff
@@ -0,0 +1,57 @@
+
+@Entity
+class Cart(
+    @OneToMany(mappedBy = "cart", cascade = [CascadeType.ALL], orphanRemoval = true)
+    val items: MutableList<CartProduct> = mutableListOf(),
```
I see you've used bidirectional mapping between `Cart` and `CartProduct`!
I'm curious about your experience in maintaining this structure.
What were the pros and cons?

I ask this because a lot of developers have different thoughts on how to manage this relationship.

(I'm not pointing out that this is wrong, so please share your thoughts without pressure 🙂 )

AnnLvu replied:
Initially, we kept the cart structure unidirectional - `CartProduct` just had a `cart_id`, and `Cart` itself didn’t hold a collection of items.
It worked, but it meant that any cart operations required explicitly working with the `CartProduct `repository. 
Later, we switched to the bidirectional mapping you see now:
```
// That's  Cart
@OneToMany(mappedBy = "cart", cascade = [CascadeType.ALL], orphanRemoval = true)
val items: MutableList<CartProduct> = mutableListOf()

// That's CartProduct
@ManyToOne(fetch = FetchType.LAZY)
@JoinColumn(name = "cart_id", nullable = false)
var cart: Cart
```
We changed this because:

- It allows us to manage the cart directly through `cart.items`, like in `addProduct()` and `decrementProduct()` in our `Cart` class:

```
fun addProduct(
        option: Option,
        quantity: Int = 1,
    ): CartProduct {
        val existing = findProduct(option)
        return if (existing != null) {
            existing.quantity += quantity
            existing
        } else {
            val newItem = CartProduct(this, option, quantity)
            items.add(newItem)
            newItem
        }
    }
```
- `orphanRemoval = true` automatically removes `CartProduct` rows when we clear the list,
so `CartService.clearCart()` can simply do:
```
 fun clear() {
        items.clear()
    }
```
**Trade‑off:**

We need to keep both sides in sync and be mindful of lazy loading,
but for a frequently updated cart this approach keeps the service layer clean and avoids manual repository calls.

pkeugine replied:
Perfect explanation! 💯

pkeugine commented on src/main/kotlin/ecommerce/model/Cart.kt:
```diff
@@ -0,0 +1,57 @@
+class Cart(
+    @OneToMany(mappedBy = "cart", cascade = [CascadeType.ALL], orphanRemoval = true)
+    val items: MutableList<CartProduct> = mutableListOf(),
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long = 0L,
```
So many new kinds of annotations were added as we migrated to JPA right? 🥲 
You did well adding in `@Id` and `@GeneratedValue` here.

I'm curious, what does `GenerationType.IDENTITY` do, and what are the other options?
Also, what made you choose `0L` to be the default value for id? I guess there are two things we could choose from : `null` and `0L`. If we chose null, id would be `Long?`.

AnnLvu replied:
I used `@GeneratedValue(strategy = GenerationType.IDENTITY`) because Justin showed it in the lecture, so I decided to stick with it. It basically means the database auto‑generates the id for each new row.
I also checked the other strategies:

- `AUTO` - JPA picks depending on the DB

- `SEQUENCE` - Uses a database sequence (common in Postgres/Oracle)

- `TABLE` - Uses a separate table (rarely used now)

I kept the default `id = 0L` so it’s non‑nullable, and JPA will just replace 0 with the real id on insert.
If I used `Long?` with `null`, I’d have to handle nulls everywhere.

pkeugine replied:
I personally prefer keeping the value null when there isn't a value set for that property, since I like to keep things consistent across all fields. I say this because there are times where fields don't have a good 'default value' or 'value that is great to express nothing'. However like your thought, there are developers who prefer `0L` since they try their best to keep things non-null. So they each have their pros and cons. Great explanation!

pkeugine commented on src/main/kotlin/ecommerce/model/CartProduct.kt:
```diff
@@ -0,0 +1,31 @@
+    name = "cart_products",
+    uniqueConstraints = [UniqueConstraint(columnNames = ["cart_id", "option_id"])],
+)
+class CartProduct(
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "cart_id", nullable = false)
+    var cart: Cart,
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "option_id", nullable = false)
+    var option: Option,
+    @Column(name = "quantity", nullable = false)
+    var quantity: Int = 1,
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long = 0L,
+)
```
```suggestion
class CartProduct(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0L,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "cart_id", nullable = false)
    var cart: Cart,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "option_id", nullable = false)
    var option: Option,

    @Column(name = "quantity", nullable = false)
    var quantity: Int = 1,
)
```

I personally prefer keeping `id` field on the very top, since that (to me) is the most important value when it comes to entity.

Also, when there's multiple annotations added, I like to keep some line breaks in between fields, as it helps increase the readability. (this is also just my opinion 😂 )

AnnLvu replied:
I actually keep the id field at the end for a practical reason:
in Kotlin, if the first constructor parameter has a default value (0L), you still need to use named arguments for all the following parameters when creating the object.
For example, if `id` were first, creating an `Option` in tests would look like this:
```
Option(
    id = 0L,
    name = "test",
    price = 10.2,
    quantity = 52,
    imageUrl = "http://localhost:8080/image/upload/product1.png",
)
```
By keeping `id` at the end, I can just rely on its default value and create test much more cleanly:
```
Option(
    "test",
    10.2,
    52,
    "http://localhost:8080/image/upload/product1.png",
)
```
It also feels more natural for me to read the business fields first, and the technical id last. 😄

But I get your point about readability - I’ve now added some line breaks between fields with multiple annotations to make it easier to scan visually.
Thanks for the tip!

AnnLvu replied:
Justin actually showed us this Kotlin trick in the lecture - if you move the id to the end with a default value, creating test  becomes so much cleaner.

I used to always keep `id` first, but after that lecture I went through and moved all `id`s to the end.😅

AnnLvu replied:
I use KtlinFormat to keep the code neat, and it removes empty lines. I need to use it to pass the Check validation. So, I can’t leave empty lines for separation if I want the Check to pass.🤣

pkeugine replied:
Great! Since this is a suggestion, I understand your point and keeping it as is looks fine 👍

pkeugine commented on src/main/kotlin/ecommerce/repository/CartStatisticRepository.kt:
```diff
@@ -0,0 +1,51 @@
+    fun findTopProducts(
+        @Param("action") action: CartAction,
+        @Param("since") since: LocalDateTime,
+        pageable: org.springframework.data.domain.Pageable,
```
```suggestion
        pageable: Pageable,
```

Why don't we simplify type names such as `org.springframework.data.domain.Pageable` to just `Pageable`? It would make reading the logic a lot easier!

Also, it would be interesting to find a way to reduce unnecessary parts in SQL, such as `new ecommerce.dto.cartStatistics.TopAddedProductsDTO`. There are multiple ways to do it, so feel free to look up some options and tell me what you think about them.

AnnLvu replied:
Thanks for the tips!
I went ahead and:

- Imported types like Pageable so the method signatures are shorter and easier to read
- Switched JPQL to interface projections instead of those long new `ecommerce.dto.cartStatistics.TopAddedProductsDto(...)` lines

In the service I just map the interfaces to our existing DTOs, so everything outside the repo stays the same.

It definitely feels cleaner now - thanks for pointing it out!

pkeugine commented on src/main/kotlin/ecommerce/model/Cart.kt:
```diff
@@ -0,0 +1,57 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    val id: Long = 0L,
+) {
+    fun addProduct(
+        option: Option,
+        quantity: Int = 1,
+    ): CartProduct {
+        val existing = findProduct(option)
+        return if (existing != null) {
+            existing.quantity += quantity
+            existing
+        } else {
+            val newItem = CartProduct(this, option, quantity)
+            items.add(newItem)
+            newItem
+        }
+    }
```
Great job not forgetting about this logic as these models are linked as bidirectional. 👍 

Since there are domain logics, it would be great to have unit tests for these logics.
Same applies to all the `require` logics in other models (domains)!

AnnLvu replied:
I added unit tests for the domain logic:

- Cart (addProduct, decrementProduct, clear and exception cases)

- Option (all require validations for name, price, quantity, and URL)

- Product (checks for at least one option and uniqueness of options)

This really helped me double‑check the business rules.
Thanks for the reminder!

pkeugine commented on src/main/kotlin/ecommerce/utils/extensions/OptionUtil.kt:
```diff
@@ -0,0 +1,24 @@
+        quantity,
+        imageUrl,
+    )
+}
```
I personally think the current `util` package is having too much things in it : `custom exceptions`, `dto - entity mappers`, `custom annotation that is related to Spring MVC`, and `JwtProvider`.

This would be a great time for us to think about `why do we separate packages`, and `when would we need one, while the other might not be needed`.

AnnLvu replied:
You were right - utils really turned into a “dumping ground”.
I went ahead and split it up:
exception - all custom exceptions

- mapper -  entity mappers & repository extensions

- annotation - LoginMember

- security - JwtProvider

Now the structure feels way cleaner and easier to navigate.
Thanks for nudging me to tidy it up! 🙌

pkeugine commented on src/main/kotlin/ecommerce/dto/cartProduct/CartProductResponse.kt:
```diff
@@ -0,0 +1,5 @@
+package ecommerce.dto.cartProduct
+
+class CartProductResponse(
```
I'm curious about how you manage class names, especially DTO names!
Is there your own convention?

I'm asking this question, because to me, it's not so straight forward on how to manage the names if I were to manage this project together. For example, there are names like :  
* `CartProductDTO`
* `CartProductResponse`
* `ProductResponseDTO`
* (if we look at other domain) `AuthTokenPayload`

I personally think class, method, and even variable names are extremely important, since when named well, they deliver their intention and purpose very well. If named without detail, we'll be the ones to look up their intentions every time we have to use them.

Why don't we refactor the names  so that they have some kind of convention that we can easily understand? :)
(If you had the convention but I didn't catch it, I'm sorry 🥲 )

AnnLvu replied:
You were right - our DTO names were a bit all over the place.
I went through and made them consistent:

- Requests now end with RequestDto
- Responses now end with ResponseDto
- Patch DTOs keep PatchDto
Thanks for pointing it out - it feels much cleaner now! 🙌

pkeugine commented on src/main/kotlin/ecommerce/model/Cart.kt:
```diff
@@ -0,0 +1,57 @@
+
+@Entity
+class Cart(
+    @OneToMany(mappedBy = "cart", cascade = [CascadeType.ALL], orphanRemoval = true)
```
**Q. Is it appropriate to use CascadeType.ALL in these contexts, or would it be better to specify only the required cascade types (PERSIST, MERGE, etc.) more explicitly?**

Nice question!
I would stick with `ALL` for now, since `Cart` owns the full lifecycle of `CartProduct`, and `Product` owns `Option`'s.

It would be more fun to ask 
* `so then, when would an entity not own child's full lifecycle?`
*  or `even if it did, when would we not use CascadeType.ALL?`. 

I would like to know your take on this. 😄

pkeugine commented on src/main/kotlin/ecommerce/controller/admin/AdminProductController.kt:
```diff
@@ -0,0 +1,127 @@
+    ): ResponseEntity<MessageResponse> {
+        adminProductService.updateOption(productId, optionId, optionDTO)
+        return ResponseEntity.ok(MessageResponse("Option updated"))
+    }
```
As two entities get involved, we can't just say `id` since we don't know which `id` we're referring to. That made us explicitly declare `productId` and `optionId` in our path.

Although this might seem not too important, I like to keep a convention on this. Rather declare all the id names explicitly even when only one entity is getting used, or allow us to use `id` for the entity that's closely related to the api.

How do you think about this? It's fine to not apply this review, but I'm curious about your thoughts.

AnnLvu replied:
Thanks for your answer!
I think `CascadeType.ALL` is good here because Cart really owns its` CartProducts` -
if cart deleted or cleared, items must be deleted too. Same with Product and Option.

About your question - when entity doesn’t own full lifecycle:
maybe when children can live alone or belong to other parents too.
In this case, using only `PERSIST` and `MERGE` is better to avoid deleting children by accident.

To be honest, I never used other cascade types before, only `ALL`,
and I’m not sure when exactly better to use different ones.

pkeugine replied:
I think your answer is placed in a different conversation 😂 
But your answer seems right on track 👍 
When child can live by itself or belongs to other parents, that's when we'll be thinking about using other cascade strategy. You say you've never met those cases, but soon you'll meet them (the time will come).
Until then, keeping these things in mind would definitely help.

Nice explanation! 👍

pkeugine commented on src/main/kotlin/ecommerce/service/AdminProductService.kt:
```diff
@@ -0,0 +1,182 @@
+    private val optionRepository: OptionRepository,
+) {
+    fun getAllProducts(
+        page: Int = DEFAULT_PAGE,
+        perPage: Int = PER_PAGE,
```
The service currently depends on controller-related information, which I generally try to avoid.

In a typical layered architecture, the flow of logic goes **top-down** like `controller → service → persistence`. To keep the architecture flexible and maintainable, it's best if each layer doesn't depend on the layer above it. That way, changes in the controller won’t ripple down and affect the service layer unnecessarily.

What do you think about this approach? 😄
No pressure at all, feel free to share your thoughts!

AnnLvu replied:
I created a new PaginationConstants object inside the config package to hold `DEFAULT_PAGE` and `PER_PAGE`.
Now both the controller and service use these constants, so the service doesn’t depend on the controller anymore.

pkeugine commented on src/main/kotlin/ecommerce/service/AdminProductService.kt:
```diff
@@ -0,0 +1,182 @@
+import java.net.URI
+
+@Service
+@Transactional
```
It's interesting that you've used `@Transactional` annotation!
What does this do, and what are the options we can use for this annotation?
(This question is pretty important, as it's a regular in tech interviews)

update : after finding out which `@Transactional` annotation you've added, I want you to tell me the difference between different `@Transactional` annotations!

AnnLvu replied:
I changed from jakarta.transaction.Transactional to Spring’s @Transactional because:

Spring’s annotation is designed specifically for Spring apps and works better with its tools.

It lets us control transactions more carefully - for example, how they start, how they behave if something goes wrong, and how nested transactions work.

jakarta.transaction.Transactional is simpler and part of a Java standard, but it doesn’t have as many features.

AnnLvu replied:
@Transactional is an annotation that helps manage database operations.
It means: “All the code inside this method should run as one single operation.”
If something goes wrong, all changes are undone to keep the data safe and consistent.

This is important to make sure the database stays correct and nothing breaks.

pkeugine commented on src/main/kotlin/ecommerce/service/AdminProductService.kt:
```diff
@@ -0,0 +1,182 @@
+        product.options.add(newOption)
+
+        return URI.create("/products/${product.id}/options/${newOption.id}")
+    }
```
As we introduced `Option` to our project, I think this `AdminProductService` became thick, meaning it's holding onto multiple domains and is ready to be separated. Controller feels the same too, but I feel like service is more important since it's closer to our application logic.

How do you feel about maintaining `Option`s in their own service?

AnnLvu replied:
I created a separate `AdminOptionService` to handle all Option-related operations. This way, each service has a clear single responsibility, and the code is easier to maintain and extend.

pkeugine commented on src/main/kotlin/ecommerce/service/AdminProductService.kt:
```diff
@@ -0,0 +1,182 @@
+    fun createOption(
+        productId: Long,
+        optionDTO: OptionDTO,
+    ): URI {
```
I see service is returning `URI` as a result for adding a new entity.
What if we added `OutputView` (remember this guy? 😄 ) and added support for consoles?
Let's say we don't provide HTTP related information to console application. Then, we would have to make a new function for creating entities, or update this logic whereas if this function returned something that could be used in any controller, we would just make `OutputView` and get it over with.

I think this kind of thoughts add up and allow us to make a maintainable, flexible-to-expansion projects. How do you think about this? 😄

AnnLvu replied:
I created a new DTO called `CreateOptionResult` to return meaningful data from the service instead of just URI.
Also updated the service and controller to use this new DTO.

pkeugine commented on src/main/kotlin/ecommerce/service/AdminProductService.kt:
```diff
@@ -0,0 +1,182 @@
+import ecommerce.utils.extensions.getPaginatedDTOs
+import ecommerce.utils.extensions.toEntity
+import ecommerce.utils.extensions.toProductDTO
+import jakarta.transaction.Transactional
```
Are you sure you want to import this `Transactional` instead of `org.springframework.transaction.annotation.Transactional`? 😄

AnnLvu replied:
I changed from jakarta.transaction.Transactional to Spring’s @transactional.
It was mistake :)

pkeugine commented on src/main/kotlin/ecommerce/model/Product.kt:
```diff
@@ -0,0 +1,30 @@
+
+@Entity
+class Product(
+    @Column(name = "name", nullable = false, unique = true)
+    var name: String,
+    @OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true)
+    @JoinColumn(name = "product_id")
+    var options: MutableList<Option>,
+    @Column(name = "created_at", nullable = false)
+    var createdAt: LocalDateTime = LocalDateTime.now(),
```
**Q. I'm also interested in whether I've correctly understood and applied the JPA relationships themselves (e.g., mappedBy, joincolumn, orphanRemoval, etc.). Are there any aspects of the mapping you think could be improved?**

I think this is a topic that can't be answered with just one comment.
I'm not saying there's a lot of wrong implementations here, what I'm trying to say is that this mission is made to let you think and experience it yourself :)

You've done a great job and I'll be sure to alert you if things feel like they need directions 👍

pkeugine reviewed (changes requested):
Hello Ann!
Learning Spring Boot is hard already, but learning JPA as well?
That requires intense focus, and you did well on it 👍

I left some reviews, questions, and answers to your questions.
Feel free to request for another review once you're done with all of them.
Some answers might feel vague, but I tried my best to let you experience this mission to it's fullest, rather than me giving you all the answers (they may not even be an answer anyways since many developers have their own point of views).

Feel free to ask back any questions via DM or comments!
Happy Hacking! 🚀 🚀 🚀

AnnLvu committed:
refactor(application.properties): enable the "Ensure line feed at file end on Save"
- remove unused dependency

AnnLvu committed:
refactor: rename files *DTO -> *Dto

AnnLvu committed:
refactor: improve entity readability with line breaks

AnnLvu committed:
refactor(CartStatisticRepository): simplify type references
- replace `org.springframework.data.domain.Pageable` with Pageable

AnnLvu committed:
refactor(CartStatisticRepository): migrate JPQL constructor expressions to interface projections
- add `TopAddedProductView` for top products
- add `MembersWhoAddedToCartView` for active users
- update `AdminStatisticsService`

AnnLvu committed:
test(model): add unit tests for Cart, Option, and Product logic
- add tests for Cart:
    • addProduct
    • decrementProduct
    • clear
- add tests for Option:
    • all require validations for name, price, quantity, and image URL
- add tests for Product:
    • requires at least one option
    • options must be distinct

AnnLvu committed:
refactor(exception): move custom exceptions out of utils

AnnLvu committed:
refactor(utils): remove utils package and reorganize files

AnnLvu committed:
refactor(dto): unify DTO naming and reorganize packages

AnnLvu committed:
refactor(config): create PaginationConstants and replace controller constants

AnnLvu committed:
refactor: replace jakarta.transaction.Transactional with Spring’s @Transactional

AnnLvu committed:
refactor(AdminProductService): move option-related logic from AdminProductService to AdminOptionService

AnnLvu committed:
refactor(Option): create new folder Options

AnnLvu committed:
feat: add CreateOptionResult DTO to encapsulate option creation response

AnnLvu committed:
refactor: use ktlintFormat

pkeugine reviewed (approved):
Hello Ann!
You did a great job refactoring this project, and all the answers you've left to my questions look great 😮 👍 

I think we are ready to move onto the next step!
I added some reactions and my thoughts to your answers so check them out when you have the time. Honestly, the answers were very solid so there weren't much to add.

If you have any questions or thoughts, feel free to DM me anytime :)
We're almost at the end of this level!
Happy Hacking! 👍 👍 👍
